firmware/applications/SignBuddy/ble_uart.c

```diff
@@ -29,22 +39,65 @@ static void uart_init(void)
   uart_config.HardwareFlowControl = BLE_UART_FLOWCTRL;
   BLE_UART_Init(BLE_UART, &uart_config);
   BLE_UART_Enable(BLE_UART);
+
+  BLE_UART_EnableIT_RXNE(BLE_UART);
+  BLE_UART_EnableIT_TC(BLE_UART);
 }
 
-void BLE_UART_IRQHandler(void)
-{}
+static inline void tx(void)
+{
+  uint8_t write_byte;
+
+  if (fifo_pop(&tx_fifo, &write_byte, 1) == 1) {
+    BLE_UART_TX(BLE_UART, write_byte);
+    writing = true;
+  }
+}
 
-uint8_t ble_uart_rx(void)
+bool ble_uart_is_writing(void)
 {
-  return BLE_UART_RX(BLE_UART);
+  return writing;
 }
 
-void ble_uart_tx(uint8_t data)
+void ble_uart_write(uint8_t *data, uint32_t length)
 {
-  BLE_UART_TX(BLE_UART, data);
+  DISABLE_IRQ();
+  fifo_push(&tx_fifo, data, length);
+  if (!writing) {
+    tx();
+  }
+  ENABLE_IRQ();
+}
+
+int ble_uart_read(uint8_t *data, uint32_t length)
+{
+  DISABLE_IRQ();
+  int ret = fifo_pop(&rx_fifo, data, length);
+  ENABLE_IRQ();
+  return ret;
 }
 
 void ble_uart_init(void)
 {
   uart_init();
+  fifo_init(&rx_fifo, rx_buffer, BLE_UART_RX_BUFFER_SIZE);
+  fifo_init(&tx_fifo, tx_buffer, BLE_UART_TX_BUFFER_SIZE);
+}
+
+void BLE_UART_IRQHandler(void)
+{
+  DISABLE_IRQ();
+  uint8_t data;
+  if (BLE_UART_IsActiveFlag_RXNE(BLE_UART)) {
+    data = BLE_UART_RX(BLE_UART);
+    // TODO: make sure this is correct flag to clear
+    BLE_UART_ClearFlag_NE(BLE_UART);
+    fifo_push(&rx_fifo, &data, 1);
+  }
+  if (BLE_UART_IsActiveFlag_TC(BLE_UART)) {
+    writing = false;
+    BLE_UART_ClearFlag_TC(BLE_UART);
+    tx();
+  }
+  ENABLE_IRQ();
```

Comment:
did you check if this irq handling works for the lpuart too? I'm not too sure about the differences between lpuart and usart

Reply:
Didn't explicitly test this interrupt. I tried receiving from the uart app and I couldn't seem to receive anything